rfcs/20190115-dense-attention.md

```diff
@@ -0,0 +1,506 @@
+  2. Calculate scores with shape `[batch_size, Tq, Tv]` as a non-linear
+     sum: `scores = tf.reduce_sum(tf.tanh(query + value), axis=-1)`
+  3. Use scores to calculate a distribution with shape
+     `[batch_size, Tq, Tv]`: `distribution = tf.nn.softmax(scores)`.
+  4. Use `distribution` to create a linear combination of `value` with
+     shape `batch_size, Tq, dim]`:
+     `return tf.matmul(distribution, value)`.
+
+  Args:
+    normalize: If True, will create scale and bias variables to normalize
+    scores.
+  """
+
+  def __init__(
+      self,
+      normalize=False,
+      **kwargs):
+
+  def build(self, input_shape):
+    """Creates variables."""
+
+  def call(self, inputs, mask=None):
+    """Applies additive attention.
+
+    Args:
+      inputs: List of the following tensors:
+        * query: Query `Tensor` of shape `[batch_size, Tq, dim]`.
+        * value: Value `Tensor` of shape `[batch_size, Tv, dim]`.
+      mask: List of the following tensors:
+        * query_mask: A boolean mask `Tensor` of shape `[batch_size, Tq]`.
+          If given, the output will be zero at the positions where
+          `mask==False`.
+        * value_mask: A boolean mask `Tensor` of shape `[batch_size, Tv]`.
+          If given, will apply the mask such that values at positions where
+          `mask==False` do not contribute to the result.
+    Returns:
+      Attention outputs of shape `[batch_size, Tq, dim]`.
+    """
+```
+
+The implementations for both Attention layers can be in the same file. They can
+reuse a private method with the following signature:
+
+```python
+def _apply_attention_scores(scores, value, value_mask=None):
+  """Applies attention scores to the given value tensor.
+
+  Args:
+    scores: Scores tensor of shape `[batch_size, Tq, Tv]`.
+    value: Value tensor of shape `[batch_size, Tv, dim]`.
+    value_mask: A boolean mask `Tensor` of shape `[batch_size, Tv]`.
+      If given, will apply the mask such that values at positions where
+      `mask==False` do not contribute to the result.
+
+  Returns:
+    Tensor of shape `[batch_size, Tq, dim]`.
+  """
+```
+
+Implementations of other Attention mechanisms can reuse this method, as well.
+So, that method can be made public. An alternative using inheritance is
+discussed in the "Base Attention Class" section. The
+rest of the code is specific to each Attention mechanism.
+
+Although not the primary focus of this proposal, the Attention layers work with
+RNN networks, such as
+[tf.keras.layers.LSTM](https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTM):
+When creating the LSTM, users need to set `return_sequences=True`, and the rest
+works the same way as CNN. It is unclear whether this method suffices to create
+the most common RNN+Attention models.
+
+We will first work on the implementation for Tensorflow.
+
+### Self-Attention
+
+The Self-Attention variant can be implemented by passing the same tensor to both
+`query` and `value`.
+
+### Multi-Head Attention
```

Comment:
I think this is covered by your proposal; just use multiple Attention layers; each will represent one attention head.

Reply:
Good point. It is not very pretty, but it works. Let me update the text.

Reply:
Multi head attention is usually batched so I think it requires a special case.

Reply:
All tensors are assumed to be batched. Namely, input tensors are `[batch_size, dim]` etc. Is this what you mean?

Reply:
I meant multi head attention is usually a single layer that takes `[batch_size, num_heads, time, dim]` and not `num_heads` layers that take `[batch_size, time, dim]`.

Reply:
Well, then the user needs to split the tensor into `num_heads` tensors of shape `[batch_size, time, dim]`. This is one way to do it.